sound/soc/sof/intel/hda.c

```diff
@@ -45,55 +45,49 @@
  * Helper function to set up a DAI widget in the DSP and then send the updated DAI config during
  * hw_params or send the updated DAI_CONFIG and then free the DAI widget during hw_free
  */
-int hda_ctrl_dai_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_dai *sof_dai,
-			      struct sof_ipc_dai_config *config, bool setup, int dir)
+int hda_ctrl_dai_widget_setup(struct snd_soc_dapm_widget *w, bool setup)
 {
-	struct snd_sof_widget *swidget;
+	struct snd_sof_widget *swidget = w->dobj.private;
+	struct snd_soc_component *component = swidget->scomp;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct sof_ipc_dai_config *config;
+	struct snd_sof_dai *sof_dai;
 	struct sof_ipc_reply reply;
-	bool found = false;
 	int ret;
 
-	/* find snd_sof_widget of type DAI with matching sname as that of snd_sof_dai name */
-	list_for_each_entry(swidget, &sdev->widget_list, list) {
-		if (dir == SNDRV_PCM_STREAM_PLAYBACK && swidget->id != snd_soc_dapm_dai_in)
-			continue;
-
-		if (dir == SNDRV_PCM_STREAM_CAPTURE && swidget->id != snd_soc_dapm_dai_out)
-			continue;
+	sof_dai = swidget->private;
 
-		if (!strcmp(sof_dai->name, swidget->widget->sname)) {
-			found = true;
-			break;
-		}
-	}
-
-	if (!found) {
-		dev_err(sdev->dev, "error: failed to find DAI widget for: %s\n",
-			sof_dai->cpu_dai_name);
+	if (!sof_dai || !sof_dai->dai_config) {
+		dev_err(sdev->dev, "No config for DAI %s\n", w->name);
 		return -EINVAL;
 	}
 
+	config = &sof_dai->dai_config[sof_dai->current_config];
+
 	/*
 	 * For static pipelines, the DAI widget would already be set up and calling
 	 * sof_widget_setup()/free() simply returns without doing anything.
 	 * For dynamic pipelines, the DAI widget will be set up or freed here.
 	 */
 	if (setup) {
-		/* set up widget first */
 		ret = sof_widget_setup(sdev, swidget);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: setting up DAI widget %s\n", w->name);
+			return ret;
+		}
```

Comment:
the return value was ignored before this patch? So, it's a separate fix?

Reply:
It's a fixup to the patch that added the function so I'd say ok to bake into this one.